internal/attestation/tdx/validator.go

```diff
@@ -0,0 +1,170 @@
+	"github.com/google/go-tdx-guest/verify/trust"
+	"github.com/prometheus/client_golang/prometheus"
+	"google.golang.org/protobuf/proto"
+)
+
+// Even though the vendored file has "SGX" in its name, it is the general "Provisioning Certificate for ECDSA Attestation"
+// from Intel and used for both SGX *and* TDX.
+//
+// See https://api.portal.trustedservices.intel.com/content/documentation.html#pcs for more information.
+//
+// File Source: https://certificates.trustedservices.intel.com/Intel_SGX_Provisioning_Certification_RootCA.pem
+//
+//go:embed Intel_SGX_Provisioning_Certification_RootCA.pem
+var tdxRootCert []byte
+
+// Validator validates attestation statements.
+type Validator struct {
+	validateOptsGen validateOptsGenerator
+	callbackers     []validateCallbacker
+	certGetter      trust.HTTPSGetter
+	logger          *slog.Logger
+	metrics         metrics
+}
+
+type metrics struct {
+	attestationFailures prometheus.Counter
+}
+
+type validateCallbacker interface {
+	ValidateCallback(ctx context.Context, quote *tdx.QuoteV4, validatorOID asn1.ObjectIdentifier,
+		reportRaw, nonce, peerPublicKey []byte) error
+}
+
+type validateOptsGenerator interface {
+	TDXValidateOpts(report *tdx.QuoteV4) (*validate.Options, error)
+}
+
+// StaticValidateOptsGenerator returns validate.Options generator that returns
+// static validation options.
+type StaticValidateOptsGenerator struct {
+	Opts *validate.Options
+}
+
+// SNPValidateOpts return the TDX validation options.
+func (v *StaticValidateOptsGenerator) TDXValidateOpts(_ *tdx.QuoteV4) (*validate.Options, error) {
+	return v.Opts, nil
+}
+
+// NewValidator returns a new Validator.
+func NewValidator(optsGen validateOptsGenerator, certGetter trust.HTTPSGetter, log *slog.Logger) *Validator {
+	return &Validator{
+		validateOptsGen: optsGen,
+		certGetter:      certGetter,
+		logger:          log,
+	}
+}
+
+// NewValidatorWithCallbacks returns a new Validator with callbacks.
+func NewValidatorWithCallbacks(optsGen validateOptsGenerator, certGetter trust.HTTPSGetter, log *slog.Logger, attestationFailures prometheus.Counter, callbacks ...validateCallbacker) *Validator {
+	v := NewValidator(optsGen, certGetter, log)
+	v.callbackers = callbacks
+	v.metrics = metrics{attestationFailures: attestationFailures}
+	return v
+}
+
+// OID returns the OID of the validator.
+func (v *Validator) OID() asn1.ObjectIdentifier {
+	return oid.RawTDXReport
+}
+
+// Validate a TDX attestation.
+func (v *Validator) Validate(ctx context.Context, attDocRaw []byte, nonce []byte, peerPublicKey []byte) (err error) {
+	v.logger.Info("Validate called", "nonce", hex.EncodeToString(nonce))
+	defer func() {
+		if err != nil {
+			v.logger.Error("Failed to validate attestation document", "err", err)
+			if v.metrics.attestationFailures != nil {
+				v.metrics.attestationFailures.Inc()
+			}
+		}
+	}()
+
+	// Parse the attestation document.
+
+	quote := &tdx.QuoteV4{}
+	if err := proto.Unmarshal(attDocRaw, quote); err != nil {
```

Comment:
Why does the Issuer return a marshalled proto quote, instead of just the raw quote?

Reply:
I'm not sure actually. I think the raw report should even be shorter, as it's non-self-describing too